// @ui5/webcomponents-react/wrappers/enums/IllustrationMessageType.js
// Generated file - do not change manually!
/**
 * Different illustration types of Illustrated Message.
 */
export var IllustrationMessageType;
(function (IllustrationMessageType) {
    /**
     * "Add Column" illustration type.
     */
    IllustrationMessageType["AddColumn"] = "AddColumn";
    /**
     * "Add People" illustration type.
     */
    IllustrationMessageType["AddPeople"] = "AddPeople";
    /**
     * "Balloon Sky" illustration type.
     */
    IllustrationMessageType["BalloonSky"] = "BalloonSky";
    /**
     * "Before Search" illustration type.
     */
    IllustrationMessageType["BeforeSearch"] = "BeforeSearch";
    /**
     * "Connection" illustration type.
     */
    IllustrationMessageType["Connection"] = "Connection";
    /**
     * "Empty Calendar" illustration type.
     */
    IllustrationMessageType["EmptyCalendar"] = "EmptyCalendar";
    /**
     * "Empty List" illustration type.
     */
    IllustrationMessageType["EmptyList"] = "EmptyList";
    /**
     * "Empty Planning Calendar" illustration type.
     */
    IllustrationMessageType["EmptyPlanningCalendar"] = "EmptyPlanningCalendar";
    /**
     * "Error Screen" illustration type.
     */
    IllustrationMessageType["ErrorScreen"] = "ErrorScreen";
    /**
     * "Filter Table" illustration type.
     */
    IllustrationMessageType["FilterTable"] = "FilterTable";
    /**
     * "Group Table" illustration type.
     */
    IllustrationMessageType["GroupTable"] = "GroupTable";
    /**
     * "No Activities" illustration type.
     */
    IllustrationMessageType["NoActivities"] = "NoActivities";
    /**
     * "No Data" illustration type.
     */
    IllustrationMessageType["NoData"] = "NoData";
    /**
     * "No Entries" illustration type.
     */
    IllustrationMessageType["NoEntries"] = "NoEntries";
    /**
     * "No Filter Results" illustration type.
     */
    IllustrationMessageType["NoFilterResults"] = "NoFilterResults";
    /**
     * "No Email" illustration type.
     */
    IllustrationMessageType["NoMail"] = "NoMail";
    /**
     * "No Email v1" illustration type.
     */
    IllustrationMessageType["NoMail_v1"] = "NoMail_v1";
    /**
     * "No Notifications" illustration type.
     */
    IllustrationMessageType["NoNotifications"] = "NoNotifications";
    /**
     * "No Saved Items" illustration type.
     */
    IllustrationMessageType["NoSavedItems"] = "NoSavedItems";
    /**
     * "No Saved Items v1" illustration type.
     */
    IllustrationMessageType["NoSavedItems_v1"] = "NoSavedItems_v1";
    /**
     * "No Search Results" illustration type.
     */
    IllustrationMessageType["NoSearchResults"] = "NoSearchResults";
    /**
     * "No Tasks" illustration type.
     */
    IllustrationMessageType["NoTasks"] = "NoTasks";
    /**
     * "No Tasks v1" illustration type.
     */
    IllustrationMessageType["NoTasks_v1"] = "NoTasks_v1";
    /**
     * "Page Not Found" illustration type.
     */
    IllustrationMessageType["PageNotFound"] = "PageNotFound";
    /**
     * "Reload Screen" illustration type.
     */
    IllustrationMessageType["ReloadScreen"] = "ReloadScreen";
    /**
     * "Resize Column" illustration type.
     */
    IllustrationMessageType["ResizeColumn"] = "ResizeColumn";
    /**
     * "Search Earth" illustration type.
     */
    IllustrationMessageType["SearchEarth"] = "SearchEarth";
    /**
     * "Search Folder" illustration type.
     */
    IllustrationMessageType["SearchFolder"] = "SearchFolder";
    /**
     * "Simple Balloon" illustration type.
     */
    IllustrationMessageType["SimpleBalloon"] = "SimpleBalloon";
    /**
     * "Simple Bell" illustration type.
     */
    IllustrationMessageType["SimpleBell"] = "SimpleBell";
    /**
     * "Simple Calendar" illustration type.
     */
    IllustrationMessageType["SimpleCalendar"] = "SimpleCalendar";
    /**
     * "Simple CheckMark" illustration type.
     */
    IllustrationMessageType["SimpleCheckMark"] = "SimpleCheckMark";
    /**
     * "Simple Connection" illustration type.
     */
    IllustrationMessageType["SimpleConnection"] = "SimpleConnection";
    /**
     * "Simple Empty Doc" illustration type.
     */
    IllustrationMessageType["SimpleEmptyDoc"] = "SimpleEmptyDoc";
    /**
     * "Simple Empty List" illustration type.
     */
    IllustrationMessageType["SimpleEmptyList"] = "SimpleEmptyList";
    /**
     * "Simple Error" illustration type.
     */
    IllustrationMessageType["SimpleError"] = "SimpleError";
    /**
     * "Simple Magnifier" illustration type.
     */
    IllustrationMessageType["SimpleMagnifier"] = "SimpleMagnifier";
    /**
     * "Simple Mail" illustration type.
     */
    IllustrationMessageType["SimpleMail"] = "SimpleMail";
    /**
     * "Simple No Saved Items" illustration type.
     */
    IllustrationMessageType["SimpleNoSavedItems"] = "SimpleNoSavedItems";
    /**
     * "Simple Not Found Magnifier" illustration type.
     */
    IllustrationMessageType["SimpleNotFoundMagnifier"] = "SimpleNotFoundMagnifier";
    /**
     * "Simple Reload" illustration type.
     */
    IllustrationMessageType["SimpleReload"] = "SimpleReload";
    /**
     * "Simple Task" illustration type.
     */
    IllustrationMessageType["SimpleTask"] = "SimpleTask";
    /**
     * "Sleeping Bell" illustration type.
     */
    IllustrationMessageType["SleepingBell"] = "SleepingBell";
    /**
     * "Sort Column" illustration type.
     */
    IllustrationMessageType["SortColumn"] = "SortColumn";
    /**
     * "Success Balloon" illustration type.
     */
    IllustrationMessageType["SuccessBalloon"] = "SuccessBalloon";
    /**
     * "Success CheckMark" illustration type.
     */
    IllustrationMessageType["SuccessCheckMark"] = "SuccessCheckMark";
    /**
     * "Success HighFive" illustration type.
     */
    IllustrationMessageType["SuccessHighFive"] = "SuccessHighFive";
    /**
     * "Success Screen" illustration type.
     */
    IllustrationMessageType["SuccessScreen"] = "SuccessScreen";
    /**
     * "Tent" illustration type.
     */
    IllustrationMessageType["Tent"] = "Tent";
    /**
     * "TntChartArea" illustration type.
     */
    IllustrationMessageType["TntChartArea"] = "TntChartArea";
    /**
     * "TntChartArea2" illustration type.
     */
    IllustrationMessageType["TntChartArea2"] = "TntChartArea2";
    /**
     * "TntChartBar" illustration type.
     */
    IllustrationMessageType["TntChartBar"] = "TntChartBar";
    /**
     * "TntChartBPMNFlow" illustration type.
     */
    IllustrationMessageType["TntChartBPMNFlow"] = "TntChartBPMNFlow";
    /**
     * "TntChartBullet" illustration type.
     */
    IllustrationMessageType["TntChartBullet"] = "TntChartBullet";
    /**
     * "TntChartDoughnut" illustration type.
     */
    IllustrationMessageType["TntChartDoughnut"] = "TntChartDoughnut";
    /**
     * "TntChartFlow" illustration type.
     */
    IllustrationMessageType["TntChartFlow"] = "TntChartFlow";
    /**
     * "TntChartGantt" illustration type.
     */
    IllustrationMessageType["TntChartGantt"] = "TntChartGantt";
    /**
     * "TntChartOrg" illustration type.
     */
    IllustrationMessageType["TntChartOrg"] = "TntChartOrg";
    /**
     * "TntChartPie" illustration type.
     */
    IllustrationMessageType["TntChartPie"] = "TntChartPie";
    /**
     * "TntCodePlaceholder" illustration type.
     */
    IllustrationMessageType["TntCodePlaceholder"] = "TntCodePlaceholder";
    /**
     * "TntCompany" illustration type.
     */
    IllustrationMessageType["TntCompany"] = "TntCompany";
    /**
     * "TntComponents" illustration type.
     */
    IllustrationMessageType["TntComponents"] = "TntComponents";
    /**
     * "TntExternalLink" illustration type.
     */
    IllustrationMessageType["TntExternalLink"] = "TntExternalLink";
    /**
     * "TntFaceID" illustration type.
     */
    IllustrationMessageType["TntFaceID"] = "TntFaceID";
    /**
     * "TntFingerprint" illustration type.
     */
    IllustrationMessageType["TntFingerprint"] = "TntFingerprint";
    /**
     * "TntLock" illustration type.
     */
    IllustrationMessageType["TntLock"] = "TntLock";
    /**
     * "TntMission" illustration type.
     */
    IllustrationMessageType["TntMission"] = "TntMission";
    /**
     * "TntNoApplications" illustration type.
     */
    IllustrationMessageType["TntNoApplications"] = "TntNoApplications";
    /**
     * "TntNoFlows" illustration type.
     */
    IllustrationMessageType["TntNoFlows"] = "TntNoFlows";
    /**
     * "TntNoUsers" illustration type.
     */
    IllustrationMessageType["TntNoUsers"] = "TntNoUsers";
    /**
     * "TntRadar" illustration type.
     */
    IllustrationMessageType["TntRadar"] = "TntRadar";
    /**
     * "TntSecrets" illustration type.
     */
    IllustrationMessageType["TntSecrets"] = "TntSecrets";
    /**
     * "TntServices" illustration type.
     */
    IllustrationMessageType["TntServices"] = "TntServices";
    /**
     * "TntSessionExpired" illustration type.
     */
    IllustrationMessageType["TntSessionExpired"] = "TntSessionExpired";
    /**
     * "TntSessionExpiring" illustration type.
     */
    IllustrationMessageType["TntSessionExpiring"] = "TntSessionExpiring";
    /**
     * "TntSuccess" illustration type.
     */
    IllustrationMessageType["TntSuccess"] = "TntSuccess";
    /**
     * "TntSuccessfulAuth" illustration type.
     */
    IllustrationMessageType["TntSuccessfulAuth"] = "TntSuccessfulAuth";
    /**
     * "TntSystems" illustration type.
     */
    IllustrationMessageType["TntSystems"] = "TntSystems";
    /**
     * "TntTeams" illustration type.
     */
    IllustrationMessageType["TntTeams"] = "TntTeams";
    /**
     * "TntTools" illustration type.
     */
    IllustrationMessageType["TntTools"] = "TntTools";
    /**
     * "TntUnableToLoad" illustration type.
     */
    IllustrationMessageType["TntUnableToLoad"] = "TntUnableToLoad";
    /**
     * "TntUnlock" illustration type.
     */
    IllustrationMessageType["TntUnlock"] = "TntUnlock";
    /**
     * "TntUnsuccessfulAuth" illustration type.
     */
    IllustrationMessageType["TntUnsuccessfulAuth"] = "TntUnsuccessfulAuth";
    /**
     * "TntUser2" illustration type.
     */
    IllustrationMessageType["TntUser2"] = "TntUser2";
    /**
     * "Unable To Load" illustration type.
     */
    IllustrationMessageType["UnableToLoad"] = "UnableToLoad";
    /**
     * "Unable To Load Image" illustration type.
     */
    IllustrationMessageType["UnableToLoadImage"] = "UnableToLoadImage";
    /**
     * "Unable To Upload" illustration type.
     */
    IllustrationMessageType["UnableToUpload"] = "UnableToUpload";
    /**
     * "Upload Collection" illustration type.
     */
    IllustrationMessageType["UploadCollection"] = "UploadCollection";
})(IllustrationMessageType || (IllustrationMessageType = {}));
