// @ui5/webcomponents-react/dist/enums/ListMode.js
// Generated file - do not change manually!
/**
 * Defines the type of <code>List</code>.
 */
export var ListMode;
(function (ListMode) {
    /**
     * Delete mode (only one list item can be deleted via provided delete button)
     */
    ListMode["Delete"] = "Delete";
    /**
     * Multi selection mode (more than one list item can be selected).
     */
    ListMode["MultiSelect"] = "MultiSelect";
    /**
     * Default mode (no selection).
     */
    ListMode["None"] = "None";
    /**
     * Right-positioned single selection mode (only one list item can be selected).
     */
    ListMode["SingleSelect"] = "SingleSelect";
    /**
     * Selected item is highlighted and selection is changed upon arrow navigation (only one list item can be selected - this is always the focused item).
     */
    ListMode["SingleSelectAuto"] = "SingleSelectAuto";
    /**
     * Left-positioned single selection mode (only one list item can be selected).
     */
    ListMode["SingleSelectBegin"] = "SingleSelectBegin";
    /**
     * Selected item is highlighted but no selection element is visible (only one list item can be selected).
     */
    ListMode["SingleSelectEnd"] = "SingleSelectEnd";
})(ListMode || (ListMode = {}));
