// @ui5/webcomponents-react/wrappers/components/FormGroup/index.js
import { CssSizeVariables, ThemingParameters } from '@ui5/webcomponents-react-base';
import React from 'react';
import { createUseStyles } from 'react-jss';
const useStyles = createUseStyles({
    title: {
        gridColumn: 'span 12',
        display: 'flex',
        alignItems: 'center',
        height: CssSizeVariables.sapWcrFormGroupTitleHeight,
        lineHeight: CssSizeVariables.sapWcrFormGroupTitleHeight,
        fontFamily: ThemingParameters.sapFontHeaderFamily,
        color: ThemingParameters.sapGroup_TitleTextColor,
        fontSize: ThemingParameters.sapFontHeader6Size,
        fontWeight: 'bold',
        backgroundColor: ThemingParameters.sapGroup_TitleBackground,
        margin: 0,
        marginBlockStart: '1rem'
    },
    spacer: { height: '1rem', gridColumn: 'span 12' }
}, { name: 'FormGroup' });
/**
 * The `FormGroup` encapsulates `FormItems` into groups.
 * __Note:__ `FormGroup` is only used for calculating the final layout of the `Form`, thus it doesn't accept any other props than `heading` and `children`, especially no `className`, `style` or `ref`.
 */
const FormGroup = (props) => {
    const { titleText, children } = props;
    const classes = useStyles();
    return (React.createElement(React.Fragment, null,
        React.createElement("h6", { className: classes.title, title: titleText, "aria-label": titleText, "data-component-name": "FormGroupTitle" }, titleText),
        children,
        React.createElement("span", { className: classes.spacer })));
};
FormGroup.displayName = 'FormGroup';
export { FormGroup };
