// @ui5/webcomponents-react/dist/components/AnalyticalTable/defaults/Column/Grouped.js
import iconNavDownArrow from '@ui5/webcomponents-icons/dist/navigation-down-arrow.js';
import iconNavRightArrow from '@ui5/webcomponents-icons/dist/navigation-right-arrow.js';
import React from 'react';
import { TextAlign } from '../../../../enums';
import { Icon } from '../../../../webComponents';
const tableGroupExpandCollapseIcon = {
    color: 'var(--sapContent_IconColor)',
    height: '0.75rem',
    width: '0.75rem',
    padding: '0.625rem',
    display: 'block'
};
export const Grouped = (props) => {
    const { cell, row } = props;
    const style = {};
    if (cell.column.hAlign && (cell.column.hAlign !== TextAlign.Left || cell.column.hAlign !== TextAlign.Begin)) {
        style.marginRight = 'auto';
    }
    return (React.createElement(React.Fragment, null,
        React.createElement("span", Object.assign({}, row.getToggleRowExpandedProps({ style })),
            React.createElement(Icon, { name: row.isExpanded ? iconNavDownArrow : iconNavRightArrow, style: tableGroupExpandCollapseIcon })),
        cell.render('Cell')));
};
