// @ui5/webcomponents-react/dist/themes/CustomVariables.js
export var CustomVariables;
(function (CustomVariables) {
    CustomVariables["ObjectPageSectionTitleFontFamily"] = "--ui5wcr_ObjectPage_SectionTitleFontFamily";
    CustomVariables["ObjectPageSectionTitleTextDecoration"] = "--ui5wcr_ObjectPage_SectionTitleTextDecoration";
    CustomVariables["ObjectPageSectionBorderTop"] = "--ui5wcr_ObjectPage_SectionBorderTop";
    CustomVariables["ObjectPageSubSectionBackgroundColor"] = "--ui5wcr_ObjectPage_SubSectionBackgroundColor";
    CustomVariables["ObjectPageSubSectionBorderRadius"] = "--ui5wcr_ObjectPage_SubSectionBorderRadius";
    CustomVariables["ActionSheetTopBorderRadius"] = "--ui5wcr_ActionSheet_TopBorderRadius";
    CustomVariables["ActionSheetBoxShadow"] = "--ui5wcr_ActionSheet_BoxShadow";
    CustomVariables["ActionSheetHeaderBoxShadow"] = "--ui5wcr_ActionSheet_HeaderBoxShadow";
    CustomVariables["ActionSheetMobileHeaderBackground"] = "--ui5wcr_ActionSheet_MobileHeaderBackground";
    CustomVariables["ActionSheetMobileHeaderBoxShadow"] = "--ui5wcr_ActionSheet_MobileHeaderBoxShadow";
    CustomVariables["ActionSheetMobileHeaderBorderRadius"] = "--ui5wcr_ActionSheet_MobileHeaderBorderRadius";
    CustomVariables["ActionSheetMobileHeaderTextColor"] = "--ui5wcr_ActionSheet_MobileHeaderTextColor";
    CustomVariables["SplitterContentBorderColor"] = "--ui5wcr_Splitter_ContentBorderColor";
    CustomVariables["SplitterIconColor"] = "--ui5wcr_Splitter_IconColor";
    CustomVariables["SplitterBarBorderStyle"] = "--ui5wcr_Splitter_BarBorderStyle";
    CustomVariables["SplitterBarBorderFocus"] = "--ui5wcr_Splitter_BarBorderFocus";
    CustomVariables["SplitterBarOutline"] = "--ui5wcr_Splitter_BarOutline";
    CustomVariables["SplitterBarBorderFix"] = "--ui5wcr_Splitter_BarBorderHighContrastFix";
    CustomVariables["AnalyticalTableOuterBorderBlock"] = "--ui5wcr-AnalyticalTable-OuterBorderBlock";
    CustomVariables["AnalyticalTableOuterBorderInline"] = "--ui5wcr-AnalyticalTable-OuterBorderInline";
    CustomVariables["AnalyticalTableHeaderFontFamily"] = "--ui5wcr-AnalyticalTable-HeaderFontFamily";
    CustomVariables["AnalyticalTableHeaderBorderWidth"] = "--ui5wcr-AnalyticalTable-HeaderBorderWidth";
    CustomVariables["AnalyticalTableHeaderActiveTextColor"] = "--ui5wcr-AnalyticalTable-HeaderActiveTextColor";
})(CustomVariables || (CustomVariables = {}));
export const CustomThemingParameters = new Proxy(CustomVariables, {
    get: (target, prop) => `var(${target[prop]})`
});
